File: api/ORM/api/construction_progress.py
from fastapi import APIRouter, Query, HTTPException
from typing import Optional
from construction_progress_models import construction_progress
from tortoise.expressions import Q
from pydantic import BaseModel
from datetime import datetime

construction_progress_api = APIRouter()

class constructionprogressIn(BaseModel):
    shield_id: str
    pro_name: str
    start_time: str
    sub_code: Optional[str] = None
    interval_length: Optional[float] = None
    today_ring: Optional[float] = None
    today_schedule: Optional[float] = None
    plan_ring: Optional[float] = None
    plan_schedule: Optional[float] = None
    accumulate_ring: Optional[float] = None
    accumulate_day: Optional[int] = None
    avg_schedule: Optional[float] = None

class ConstructionProgressUpdateModel(BaseModel):
    id: int
    shield_id: str
    pro_name: str
    start_time: str
    sub_code: Optional[str] = None
    interval_length: Optional[float] = None
    today_ring: Optional[float] = None
    today_schedule: Optional[float] = None
    plan_ring: Optional[float] = None
    plan_schedule: Optional[float] = None
    accumulate_ring: Optional[float] = None
    accumulate_day: Optional[int] = None
    avg_schedule: Optional[float] = None

@construction_progress_api.post("/progress")
async def create_construction_progress(progress_data: constructionprogressIn):
    try:
        data = progress_data.dict(exclude_unset=True)
        if "start_time" in data:
            try:
                data["start_time"] = datetime.strptime(data["start_time"], "%Y-%m-%d").date()
            except ValueError as e:
                raise HTTPException(status_code=400, detail=f"日期格式错误，请使用YYYY-MM-DD格式: {str(e)}")
        await construction_progress.create(**data)
        return {"status": 0, "msg": "添加成功"}
    except HTTPException as e:
        raise e
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"添加失败: {str(e)}")

@construction_progress_api.get("/progress_list")
async def get_construction_progress_list(
    shield_id: Optional[str] = Query(None, description="盾构机编号"),
    pro_name: Optional[str] = Query(None, description="区间名称"),
    sub_code: Optional[str] = Query(None, description="施工区间编码"),
    page: int = Query(1, ge=1, description="页码"),
    per_page: int = Query(10, ge=1, le=100, description="每页数量")
):
    filters = Q()
    if shield_id:
        filters &= Q(shield_id=shield_id)
    if pro_name:
        filters &= Q(pro_name=pro_name)
    if sub_code:
        filters &= Q(sub_code=sub_code)

    total = await construction_progress.filter(filters).count()
    items = await construction_progress.filter(filters) \
        .offset((page - 1) * per_page) \
        .limit(per_page) \
        .values()

    return {
        "status": 0,
        "msg": "获取成功",
        "data": {
            "items": items,
            "total": total,
            "page": page,
            "per_page": per_page
        }
    }

@construction_progress_api.post("/update_progress")
async def update_construction_progress(progress_data: ConstructionProgressUpdateModel):
    try:
        progress = await construction_progress.get_or_none(id=progress_data.id)
        if not progress:
            return {"status": 1, "msg": f"未找到ID为{progress_data.id}的记录"}
        
        data = progress_data.dict(exclude_unset=True)
        if "start_time" in data:
            try:
                data["start_time"] = datetime.strptime(data["start_time"], "%Y-%m-%d").date()
            except ValueError as e:
                raise HTTPException(status_code=400, detail=f"日期格式错误，请使用YYYY-MM-DD格式: {str(e)}")
        
        await progress.update_from_dict(data)
        await progress.save()
        return {"status": 0, "msg": "更新成功"}
    except HTTPException as e:
        raise e
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"更新失败: {str(e)}")

@construction_progress_api.delete("/delete_progress/{id}")
async def delete_construction_progress(id: int):
    try:
        progress = await construction_progress.get_or_none(id=id)
        if not progress:
            return {"status": 1, "msg": f"未找到ID为{id}的记录"}
        
        await progress.delete()
        return {"status": 0, "msg": "删除成功"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"删除失败: {str(e)}")

File: pages/construction_progress.js
(function () {
  const response = {
    "type": "page",
    "body": [
      {
        "type": "button",
        "label": "新增施工进度",
        "icon": "fa fa-plus",
        "level": "primary",
        "className": "m-b-sm",
        "actionType": "dialog",
        "dialog": {
          "title": "新增施工进度",
          "body": {
            "type": "form",
            "api": {
              "method": "post",
              "url": "/construction_progress/progress",
              "messages": {
                "success": "添加成功",
                "failed": "添加失败"
              }
            },
            "body": [
              {
                "type": "input-text",
                "name": "shield_id",
                "label": "盾构机编号",
                "required": true
              },
              {
                "type": "input-text",
                "name": "pro_name",
                "label": "区间名称",
                "required": true
              },
              {
                "type": "input-text",
                "name": "interval_length",
                "label": "区间长度"
              },
              {
                "type": "input-date",
                "name": "start_time",
                "label": "始发时间",
                "required": true,
                "format": "YYYY-MM-DD"
              },
              {
                "type": "input-text",
                "name": "today_ring",
                "label": "今日进度-环"
              },
              {
                "type": "input-text",
                "name": "today_schedule",
                "label": "今日进度-m"
              },
              {
                "type": "input-text",
                "name": "plan_ring",
                "label": "计划进度-环"
              },
              {
                "type": "input-text",
                "name": "plan_schedule",
                "label": "计划进度-m"
              },
              {
                "type": "input-text",
                "name": "accumulate_ring",
                "label": "开累进度-环"
              },
              {
                "type": "input-number",
                "name": "accumulate_day",
                "label": "累计天数-天"
              },
              {
                "type": "input-text",
                "name": "avg_schedule",
                "label": "平均日进度-m"
              }
            ]
          }
        }
      },
      {
        "type": "crud",
        "syncLocation": false,
        "resizable": true,
        "columnsTogglable": true,
        "className": "table-db m-b-none",
        "tableClassName": "table-cell-nowrap",
        "api": {
          "url": "/construction_progress/progress_list",
          "method": "get",
          "dataType": "json",
          "adaptor": function (payload) {
            return {
              status: payload.status,
              msg: payload.msg,
              data: {
                items: payload.data && payload.data.items ? payload.data.items : [],
                total: payload.data && payload.data.total ? payload.data.total : 0
              }
            };
          }
        },
        "headerToolbar": [
          {
            "type": "tpl",
            "tpl": "<div class='header-title'><i class='fa fa-list'></i> 施工进度报表</div>"
          },
          "export-excel",
          {
            "type": "reload",
            "icon": "fa fa-refresh",
            "tooltip": "刷新数据",
            "tooltipPlacement": "top",
            "level": "link"
          }
        ],
        "filter": {
          "title": "查询条件",
          "submitText": "查询",
          "body": [
            {
              "type": "input-text",
              "name": "shield_id",
              "label": "盾构机编号",
              "placeholder": "请输入盾构机编号进行查询"
            },
            {
              "type": "input-text",
              "name": "pro_name",
              "label": "区间名称",
              "placeholder": "请输入区间名称进行查询"
            }
          ]
        },
        "columns": [
          {
            "type": "operation",
            "label": "操作",
            "width": 80,
            "fixed": "right",
            "buttons": [
              {
                "type": "button",
                "icon": "fa fa-pencil",
                "tooltip": "修改",
                "level": "link",
                "actionType": "drawer",
                "drawer": {
                  "position": "right",
                  "size": "md",
                  "title": "修改施工进度",
                  "body": {
                    "type": "form",
                    "api": {
                      "method": "post",
                      "url": "/construction_progress/update_progress",
                      "messages": {
                        "success": "更新成功",
                        "failed": "更新失败"
                      }
                    },
                    "body": [
                      {"type": "hidden", "name": "id"},
                      {"type": "input-text", "name": "shield_id", "label": "盾构机编号", "required": true},
                      {"type": "input-text", "name": "pro_name", "label": "区间名称", "required": true},
                      {"type": "input-text", "name": "interval_length", "label": "区间长度"},
                      {"type": "input-date", "name": "start_time", "label": "始发时间", "required": true, "format": "YYYY-MM-DD"},
                      {"type": "input-text", "name": "today_ring", "label": "今日进度-环"},
                      {"type": "input-text", "name": "today_schedule", "label": "今日进度-m"},
                      {"type": "input-text", "name": "plan_ring", "label": "计划进度-环"},
                      {"type": "input-text", "name": "plan_schedule", "label": "计划进度-m"},
                      {"type": "input-text", "name": "accumulate_ring", "label": "开累进度-环"},
                      {"type": "input-number", "name": "accumulate_day", "label": "累计天数-天"},
                      {"type": "input-text", "name": "avg_schedule", "label": "平均日进度-m"}
                    ]
                  }
                }
              },
              {
                "type": "button",
                "icon": "fa fa-times text-danger",
                "actionType": "ajax",
                "tooltip": "删除",
                "confirmText": "确定要删除该记录吗？",
                "api": {
            "method": "delete",
            "url": "/construction_progress/delete_progress/${id}"
          },
                "messages": {
                  "success": "删除成功",
                  "failed": "删除失败"
                }
              }
            ]
          },
          { "name": "id","label": "编号", "type": "number", "width": 80 },
          { "name": "shield_id", "label": "盾构机编号", "type": "text", "width": 120 },
          { "name": "pro_name", "label": "区间名称", "type": "text", "width": 150 },
          { "name": "interval_length", "label": "区间长度", "type": "text", "width": 100 },
          { "name": "start_time", "label": "始发时间", "type": "date", "width": 120 },
          { "name": "today_ring", "label": "今日进度-环", "type": "text", "width": 100 },
          { "name": "today_schedule", "label": "今日进度-m", "type": "text", "width": 100 },
          { "name": "plan_ring", "label": "计划进度-环", "type": "text", "width": 100 },
          { "name": "plan_schedule", "label": "计划进度-m", "type": "text", "width": 100 },
          { "name": "accumulate_ring", "label": "开累进度-环", "type": "text", "width": 100 },
          { "name": "accumulate_day", "label": "累计天数-天", "type": "number", "width": 100 },
          { "name": "avg_schedule", "label": "平均日进度-m", "type": "text", "width": 100 }
        ]
      }
    ]
  };

  window.jsonpCallback && window.jsonpCallback(response);
})();